### PR TITLE
kubeadm: pod-cidr, service-cidr should not go through any proxy

### LINKED ISSUE #666: Proxy environment variables shouldn't be passed to apiserver
## What keywords did you search in kubeadm issues before filing this one?
environment, proxy

## Is this a BUG REPORT or FEATURE REQUEST?
BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`):
kubeadm version: &version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T09:42:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T09:42:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

- **Cloud provider or hardware configuration**:

- **OS** (e.g. from /etc/os-release):
NAME="Red Hat Enterprise Linux Workstation"
VERSION="7.4 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Workstation"
VARIANT_ID="workstation"
VERSION_ID="7.4"
PRETTY_NAME="Red Hat Enterprise Linux"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.4:GA:workstation"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.4
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.4"

- **Kernel** (e.g. `uname -a`):
Linux [redacted] 3.10.0-693.2.2.el7.x86_64 #1 SMP Sat Sep 9 03:55:24 EDT 2017 x86_64 x86_64 x86_64 GNU/Linux

- **Others**:

## What happened?
My cluster sits behind a corporate firewall and as such, we have http_proxy, https_proxy, and no_proxy environment variables defined in order to reach outside severs.  When calling kubdadm the enviromnment gets passed along to the apiserver manifest (and to others).  However, unless your no_proxy settings are configured to include all of the IP adresses that will be used by kubernetes for pods and services, the api server may incorrectly try to use the proxy settings to access services. 

For example, if you start up the dashboard and attempt to access it through the api server through http://localhost:8001/api/v1/namespaces/kube-system/services/https:kubernetes-dashboard:/proxy/, the service will fail with an EOF error because the api server cannot access the dashboard (because it attempts to use the proxy settings).  

## What you expected to happen?
To the best of my knowledge, the api server never needs to access the outside world, only local services and endpoints, therefore, I see no reason why the environment should be passed to it when the manifest is created.

## How to reproduce it (as minimally and precisely as possible)?
[ See example above ]

## Anything else we need to know?
The obvious workaround is to not have proxy settings defined when you call kubeadm init.  And this is a workable solution.  However, some of the services that start up do expect to call home for various reasons (e.g., kubeadm attempts to fetch the latest version number of kubernetes if it's not passed as a command line), and those requests will fail if required proxy settings are not defined.

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
kubeadm passes proxy variables to static pods during init stage by #37494. But all the pods and service ips in pod-cidr, service-cidr should not go through any proxy.

This PR fixes this issue.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#666

**Special notes for your reviewer**:
/assign @luxas 
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm

**Release note**:

```release-note
kubeadm: pod-cidr, service-cidr should not go through any proxy. Bypassing proxies for intra-cluster communication.
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -277,21 +277,34 @@ func getSchedulerCommand(cfg *kubeadmapi.MasterConfiguration) []string {
...
 			// malformed environment variable, skip it.
 			continue
 		}
-		name := env[:pos]
+		name := strings.ToLower(env[:pos])
 		value := env[pos+1:]
-		if strings.HasSuffix(strings.ToLower(name), "_proxy") && value != "" {
+
+		if strings.HasSuffix(name, "_proxy") && value != "" {
+			if name == "no_proxy" {
```
Comment (jamiehannaford) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: you could also `continue` if name == no_proxy and set the env in 1 place rather than 2. WDYT? not a big issue IMO

**Review by jamiehannaford — COMMENTED** [review]
thanks for fixing this, just one minor nit

**Comment by kad** [comment]
First of all, kubeadm produce warnings if connections to any used CIDRs will end-up via proxy.
Altering or ignoring user set environments some time ago was considered as not appropriate solution, as in some corner cases it might be needed by cluster administrator.

As for PR itself, it handles only IPv4 addresses, that is not acceptable, as since 1.9.x release kubernetes can be both IPv4 and IPv6.

/assign kad

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -277,21 +277,34 @@ func getSchedulerCommand(cfg *kubeadmapi.MasterConfiguration) []string {
 }
 
 // getProxyEnvVars builds a list of environment variables to use in the control plane containers in order to use the right proxy
-func getProxyEnvVars() []v1.EnvVar {
+func getProxyEnvVars(cfg *kubeadmapi.MasterConfiguration) []v1.EnvVar {
+	noProxyIPs := getNoProxyIPv4IPs(cfg)
```
Comment (kad) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: kubeadm must be able to handle both IPv4 and IPv6.

**Review by kad — CHANGES_REQUESTED** [review]

**Comment by errordeveloper** [comment]
The release note is not very clear at all. This is about bypassing proxies for intra-cluster communication, right? If so, please say something like that, so that it makes sense to end-users. To be honest it took me a while to get what it is about exactly. Also, as far as I am aware, the communication of all control plane components should not even leave the network, accept for fetching images from a public registry, or check version (like kubeadm and weave net do), so I am not entirely sure why we have it set everywhere in the first place...

**Comment by kad** [comment]
@errordeveloper It is related to issues kubernetes/kubernetes#36573  kubernetes/kubeadm#5

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (dixudx) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @kad Thanks for reminding. This function works both for IPv4 and IPv6. Locally tested.

**Comment by dixudx** [comment]
@kad @jamiehannaford Updated. PTAL. Thanks.

**Comment by dixudx** [comment]
> Also, as far as I am aware, the communication of all control plane components should not even leave the network, accept for fetching images from a public registry, or check version (like kubeadm and weave net do), so I am not entirely sure why we have it set everywhere in the first place...

@kad Related to issue kubernetes/kubeadm#666 as well.

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 3]
```diff
@@ -277,21 +277,32 @@ func getSchedulerCommand(cfg *kubeadmapi.MasterConfiguration) []string {
...
+				noProxyIPs = append(noProxyIPs, value)
+				continue
+			}
 			envVar := v1.EnvVar{Name: name, Value: value}
 			envs = append(envs, envVar)
 		}
 	}
+
+	// set no_proxy
+	envVar := v1.EnvVar{Name: "no_proxy", Value: strings.Join(noProxyIPs, ",")}
```
Comment (roberthbailey) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: This could be a really long list if you have a large CIDR block. Isn't there a limit on how long an environment variable can be (less than 32k)?

Comment (kad) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: There is limit. And transport connections supports CIDR notation of subnets in no_proxy.
There is no need to expand them at all.

Comment (dixudx) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @kad AFAIK, golang does not support no_proxy with CIDR notation of subnet. 
Here is an on-going issue golang/go#16704 on this.

> Isn't there a limit on how long an environment variable can be (less than 32k)
> There is limit.

Yeah. https://www.in-ulm.de/~mascheck/various/argmax/ lists Actual values for `ARG_MAX`. This is a risk when setting environments.

Comment (kad) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: That is exactly the reason why I consider that patch is incorrect:
1. kubeadm and other kubernetes components are utilizing modified http transport which supports CIDR notations.
2. kubeadm gives warnings during init stage if access to any IP address in pod or service CIDRs will be going via proxies, according to values of *_proxy variables.
3. If system admin require traffic to be going via proxies, he is able to ignore warning.
4. if system admin wants to have traffic going directly = values of environment variables before running kubeadm should be including pod and service cidrs.
5. with current idea of the patch behaviour becomes inconsistent: kubeadm will produce warning, but anyway will alter variables put into manifests. That is less predictable for sysadmins and will cause more issues.
6. trying to expand IPv6 pod/service cidrs will lead to OOM-kill of kubeadm.

So, to summarise: if in issue  kubernetes/kubeadm#666 warning was correctly shown to user of kubeadm but was ignored by user, we don't need this PR at all. If for some reason warning was not shown, we need to investigate more.

Comment (dixudx) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @kad Thanks for the explanation.

> And transport connections supports CIDR notation of subnets in no_proxy.
There is no need to expand them at all.

I did found the link https://github.com/kubernetes/kubernetes/blob/f31ac9e0b6e70516648970699fe04f929a7a320e/staging/src/k8s.io/apimachinery/pkg/util/net/http.go#L86-L88 .

Will update it.

**Comment by dixudx** [comment]
ping @jamiehannaford @roberthbailey @kad Updated. PTAL. Thanks.

**Comment by kad** [comment]
I think that it is better to have consistent behavior, and this PR is not needed. Unless we get more information from  kubernetes/kubeadm#666, I'd prefer to handle proxy settings as it was agreed in #35044: we have pre-flight checks and they are  providing warnings to sysadmins. Sysadmins  have choice which way is better for their particular setup.

**Comment by errordeveloper** [comment]
I agree with @kad's view on this.

**Comment by dixudx** [comment]
> we have pre-flight checks and they are providing warnings to sysadmins. Sysadmins have choice which way is better for their particular setup.

Right. But most people would just ignore those warnings.

**Comment by kad** [comment]
We can make them as errors in theory. With fine-grained error skipping, that can be an option for those who has legitimate cases to use transport calls via proxies. Not a nice solution, but at least predictable and explicit.

**Comment by kad** [comment]
Let's close this PR. If we consider that proxy checks must become errors, it can be done in separate PR.

/close

**Comment by errordeveloper** [comment]
Thanks, @kad, I agree.
